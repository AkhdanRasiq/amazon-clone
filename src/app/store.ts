import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import productReducer from '../features/productSlice'
import alertReducer from '../features/alertSlice'

export const store = configureStore({
  reducer: {
    products  : productReducer,
    alert     : alertReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
