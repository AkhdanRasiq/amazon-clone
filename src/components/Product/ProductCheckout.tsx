import React from 'react'

import { useAppDispatch } from '../../app/hooks'
import { removeProduct, addProduct } from '../../features/productSlice'

interface IProductCheckoutProps {
  a_iId       : number,
  a_strTitle  : string,
  a_iPrice    : number,
  a_iRating   : number,
  a_strImage  : string,
  a_iQty      : number  
}

function ProductCheckout({ a_iId, a_strTitle, a_iPrice, a_iRating, a_strImage, a_iQty }: IProductCheckoutProps) {
  const dispatach   = useAppDispatch()

  const onBtnAddProductClick = () => {
    dispatach(addProduct({ id: a_iId }))
  }

  const onBtnRemoveProductClick = () => {
    dispatach(removeProduct({ id: a_iId }))
  }

  const ratingRender = () => {
    const ratingElement = []
    for (let iRating = 0; iRating < a_iRating; iRating++) {
      ratingElement.push(<p key={iRating}>⭐</p>)
    }
    return ratingElement
  }

  return (
    <div className='productCheckout'>
      <div>
        <img className='productCheckoutImage' src={a_strImage} alt="product_img" />
      </div>
      <div className='productCheckoutInfo'>
        <p className='productCheckoutInfoTitle'>
          {a_strTitle}
        </p>
        <p className='productCheckoutInfoPrice'>
          <small>$</small>
          <strong>{a_iPrice}</strong>
        </p>
        <div className="productCheckoutInfoRating">
          {ratingRender().map((item) => (
            item
          ))}
        </div>
        <div className='productCheckoutAction'>
          <button id='btnRemoveProduct' onClick={() => onBtnRemoveProductClick()}>─</button>
          <p>{a_iQty}</p>
          <button id='btnAddProduct' onClick={() => onBtnAddProductClick()}>+</button>
        </div>
      </div>
    </div>
  )
}

export default ProductCheckout
