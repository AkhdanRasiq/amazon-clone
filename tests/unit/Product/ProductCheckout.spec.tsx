import React from 'react'
import { render } from "@testing-library/react"
import { store } from '../../../src/app/store';
import { Provider } from 'react-redux';

import ProductCheckout from "../../../src/components/Product/ProductCheckout"

describe("ProductCheckout.tsx", () => {
  it("Render Properly", async () => {
    const wrapper = render(
      <Provider store={store}>
        <ProductCheckout 
          a_iId={1}
          a_strTitle='The Test'
          a_iPrice={19.99}
          a_iRating={5}
          a_strImage="www.test.com"
          a_iQty={1}
        />
      </Provider>
    )
    await expect(wrapper).toMatchSnapshot()
  })
})
