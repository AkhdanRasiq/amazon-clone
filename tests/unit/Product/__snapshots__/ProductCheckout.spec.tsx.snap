// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCheckout.tsx Render Properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="productCheckout"
      >
        <div>
          <img
            alt="product_img"
            class="productCheckoutImage"
            src="www.test.com"
          />
        </div>
        <div
          class="productCheckoutInfo"
        >
          <p
            class="productCheckoutInfoTitle"
          >
            The Test
          </p>
          <p
            class="productCheckoutInfoPrice"
          >
            <small>
              $
            </small>
            <strong>
              19.99
            </strong>
          </p>
          <div
            class="productCheckoutInfoRating"
          >
            <p>
              ⭐
            </p>
            <p>
              ⭐
            </p>
            <p>
              ⭐
            </p>
            <p>
              ⭐
            </p>
            <p>
              ⭐
            </p>
          </div>
          <div
            class="productCheckoutAction"
          >
            <button>
              ─
            </button>
            <p>
              1
            </p>
            <button>
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="productCheckout"
    >
      <div>
        <img
          alt="product_img"
          class="productCheckoutImage"
          src="www.test.com"
        />
      </div>
      <div
        class="productCheckoutInfo"
      >
        <p
          class="productCheckoutInfoTitle"
        >
          The Test
        </p>
        <p
          class="productCheckoutInfoPrice"
        >
          <small>
            $
          </small>
          <strong>
            19.99
          </strong>
        </p>
        <div
          class="productCheckoutInfoRating"
        >
          <p>
            ⭐
          </p>
          <p>
            ⭐
          </p>
          <p>
            ⭐
          </p>
          <p>
            ⭐
          </p>
          <p>
            ⭐
          </p>
        </div>
        <div
          class="productCheckoutAction"
        >
          <button>
            ─
          </button>
          <p>
            1
          </p>
          <button>
            +
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
